### DISCUSSION TITLE
How to reuse await session.execute(query) result?

### DISCUSSION THREAD

**Discussion by CHuKeR** [sqlalchemy/sqlalchemy#6906] [Usage Questions] [answered]
I have code:

```
query = select(A)
result = await session.execute(query)
items = result.all()
```
And after result.all() I want to result.first or result.all or result.scalar... (For debugging, for example)
But after first result.all(), which return values, second call return nothing.

session init:
```
engine = create_async_engine(
    "postgresql+asyncpg://postgres:postgres@localhost/postgres",
)

async_session = sessionmaker(
    engine, expire_on_commit=False, class_=AsyncSession
)


async def get_session() -> AsyncSession:
    async with async_session() as session:
        yield session
```

Is there way to reuse result after first call .all() function?

**Comment by zzzeek** [comment, marked as answer]
the result is a consumable object.  if you want to make copies of it instead that each can be consumed independently, you can use [result.freeze](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=result%20freeze#sqlalchemy.ext.asyncio.AsyncResult.freeze)

```py
result = await session.execute(query)
frozen = await result.freeze()

result1 = frozen()
result1.first()

result2 = frozen()
result2.all()

# etc.
```

**Reply by CHuKeR** [reply]
> the result is a consumable object. if you want to make copies of it instead that each can be consumed independently, you can use [result.freeze](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html?highlight=result%20freeze#sqlalchemy.ext.asyncio.AsyncResult.freeze)
> 
> ```python
> result = await session.execute(query)
> frozen = await result.freeze()
> 
> result1 = frozen()
> result1.first()
> 
> result2 = frozen()
> result2.all()
> 
> # etc.
> ```

Hm, looks good, but now I need to add freeze method when I want to test something!

And it doesn't work with await, I have to call:
```
result = await session.execute(query)
frozen = result.freeze()

result1 = frozen()
result1.first()
result1.all()
```
